{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Simulating MSK-IMPACT 410 Gene Panel Mutational Spectra\n",
    "\n",
    "**Goal**: simulate an MSK-IMPACT 410 gene panel mutational spectra matrix from ICGC whole genome mutational annotation file (MAF) datasets.\n",
    "\n",
    "This notebook assumes that the notebooks— `Consuming ICGC Data.ipynb` and `Processing ICGC Data.ipynb` have already been run, and in that order. Among other things, `Processing ICGC Data.ipynb` generates a set of MAF files from the simple somatic mutation (SSM) data, generated from whole genome sequencing (WGS) analysis, which we downloaded from the ICGC Data Portal. In this notebook, we will—\n",
    "* Downsample the WGS MAF files to contain only those mutations that are covered by [the MSK-IMPACT 410 targeted gene panel](https://pubmed.ncbi.nlm.nih.gov/25801821/).\n",
    "* Convert the downsampled MSK-IMPACT 410 gene panel data into a mutational spectra matrix for mutational signature analysis.\n"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "outputs": [],
   "source": [
    "from collections import OrderedDict\n",
    "from pathlib import Path\n",
    "from typing import List\n",
    "\n",
    "from natsort import natsorted\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyfaidx"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dir_data = Path.cwd().parent / \"data\"\n",
    "dir_maf_dirs = dir_data / \"WGS_MAFs\"\n",
    "\n",
    "fpath_grch37 = dir_data / \"hg19.fa\"\n",
    "fpath_msk_impact_bed = dir_data / \"MSK-IMPACT410.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [
    "dir_panel_maf_dirs = dir_data / \"MSK_IMPACT_410_MAFs\"\n",
    "if not dir_panel_maf_dirs.exists():\n",
    "        dir_panel_maf_dirs.mkdir()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [],
   "source": [
    "msk_impact_df = pd.read_csv(fpath_msk_impact_bed, sep=\"\\t\")"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "data": {
      "text/plain": "  Chromosome  Start_Position  End_Position Hugo_Symbol    ID   SEQ_ASSAY_ID  \\\n0          9       133710831     133710915        ABL1  ABL1  MSK-IMPACT410   \n1          9       133729448     133729627        ABL1  ABL1  MSK-IMPACT410   \n2          9       133730185     133730486        ABL1  ABL1  MSK-IMPACT410   \n3          9       133738147     133738425        ABL1  ABL1  MSK-IMPACT410   \n4          9       133747513     133747603        ABL1  ABL1  MSK-IMPACT410   \n\n  Feature_Type  includeInPanel  \n0         exon            True  \n1         exon            True  \n2         exon            True  \n3         exon            True  \n4         exon            True  ",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Chromosome</th>\n      <th>Start_Position</th>\n      <th>End_Position</th>\n      <th>Hugo_Symbol</th>\n      <th>ID</th>\n      <th>SEQ_ASSAY_ID</th>\n      <th>Feature_Type</th>\n      <th>includeInPanel</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>9</td>\n      <td>133710831</td>\n      <td>133710915</td>\n      <td>ABL1</td>\n      <td>ABL1</td>\n      <td>MSK-IMPACT410</td>\n      <td>exon</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>9</td>\n      <td>133729448</td>\n      <td>133729627</td>\n      <td>ABL1</td>\n      <td>ABL1</td>\n      <td>MSK-IMPACT410</td>\n      <td>exon</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>9</td>\n      <td>133730185</td>\n      <td>133730486</td>\n      <td>ABL1</td>\n      <td>ABL1</td>\n      <td>MSK-IMPACT410</td>\n      <td>exon</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>9</td>\n      <td>133738147</td>\n      <td>133738425</td>\n      <td>ABL1</td>\n      <td>ABL1</td>\n      <td>MSK-IMPACT410</td>\n      <td>exon</td>\n      <td>True</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>9</td>\n      <td>133747513</td>\n      <td>133747603</td>\n      <td>ABL1</td>\n      <td>ABL1</td>\n      <td>MSK-IMPACT410</td>\n      <td>exon</td>\n      <td>True</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "msk_impact_df.head()"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "outputs": [],
   "source": [
    "def read_sbs_maf_file(filepath: Path) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Reads only single base substitutions from an MAF file generated from an\n",
    "    ICGC SSM dataset.\n",
    "\n",
    "    :param filepath: file path to the MAF file.\n",
    "    :return: Pandas dataframe with only single base substitutions.\n",
    "    \"\"\"\n",
    "    data = pd.read_csv(filepath, sep=\"\\t\")\n",
    "    data = data.loc[data[\"mutation_type\"] == \"single base substitution\"].reset_index(drop=True)\n",
    "    return data\n",
    "\n",
    "\n",
    "def is_mutation_in_gene_panel(panel_df,\n",
    "                              chromosome: str,\n",
    "                              position_start: int,\n",
    "                              position_end: int) -> bool:\n",
    "        \"\"\"\n",
    "        Checks if the defined mutation is measured by the gene panel.\n",
    "\n",
    "        :param panel_df: gene panel BED file.\n",
    "        :param chromosome: chromosome name of the mutation.\n",
    "        :param position_start: starting position of the mutation.\n",
    "        :param position_end: end position of the mutation.\n",
    "        :return: True if the mutation region is measured in the gene panel.\n",
    "            False otherwise.\n",
    "        \"\"\"\n",
    "        panel_matches = panel_df.loc[((panel_df[\"Chromosome\"] == chromosome) &\n",
    "                                      (panel_df[\"Start_Position\"] <= position_start) &\n",
    "                                      (panel_df[\"End_Position\"] >= position_end))]\n",
    "        return len(panel_matches) > 0\n",
    "\n",
    "\n",
    "def return_panel_rows(panel_df: pd.DataFrame,\n",
    "                      wg_maf: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Takes an MAF dataframe from WGS analysis and downsamples it to only return the\n",
    "    mutations covered by the gene panel (as described by its BED file).\n",
    "\n",
    "    :param panel_df: gene panel's BED file.\n",
    "    :param wg_maf: MAF from WGS analysis.\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    panel_rows = wg_maf.apply(lambda row: is_mutation_in_gene_panel(panel_df,\n",
    "                                                                    row[\"chromosome\"],\n",
    "                                                                    row[\"chromosome_start\"],\n",
    "                                                                    row[\"chromosome_end\"]),\n",
    "                              axis=1)\n",
    "    return wg_maf.loc[panel_rows]\n",
    "\n",
    "\n",
    "def convert_wg_mafs_to_panel_mafs(dir_wg_mafs: Path,\n",
    "                                  panel_df: pd.DataFrame,\n",
    "                                  dir_panel_mafs: Path) -> None:\n",
    "        \"\"\"\n",
    "        Converts each MAF file, generated from WGS analysis, in a directory and\n",
    "        downsamples it to only MAF files containing only the mutations covered\n",
    "        by the gene panel (as described by its BED file).\n",
    "\n",
    "        :param dir_wg_mafs: a directory containing a set of MAF files from WGS\n",
    "            analysis.\n",
    "        :param panel_df: BED file of the gene panel.\n",
    "        :param dir_panel_mafs: output directory.\n",
    "        \"\"\"\n",
    "        maf_filepaths = list(dir_wg_mafs.glob(\"*\"))\n",
    "\n",
    "        for maf_filepath in maf_filepaths:\n",
    "            filepath_panel = dir_panel_mafs / f\"{maf_filepath.stem}\"\n",
    "            data_maf = read_sbs_maf_file(maf_filepath)\n",
    "            data_maf = return_panel_rows(panel_df, data_maf)\n",
    "            if len(data_maf) > 0:\n",
    "                data_maf.to_csv(filepath_panel, sep=\"\\t\", index=False)\n",
    "\n",
    "\n",
    "def panel_maf_dirs_from_wgs_maf_dirs(dir_wg_maf_dirs: Path,\n",
    "                                     panel_df: pd.DataFrame,\n",
    "                                     dir_panel_maf_dirs: Path) -> None:\n",
    "        \"\"\"\n",
    "        For each directory within the specified directory, this method iterates through\n",
    "        all MAF files, generated from WGS analysis, and downsamples it to create MAF\n",
    "        files containing mutations covered by the gene panel (as described by its BED\n",
    "        file).\n",
    "\n",
    "        :param dir_wg_maf_dirs: a directory of directories, each containing a set of MAF files\n",
    "            from WGS analysis.\n",
    "        :param panel_df: BED file of the gene panel.\n",
    "        :param dir_panel_maf_dirs: output directory.\n",
    "        \"\"\"\n",
    "        for dir_wg_mafs in dir_wg_maf_dirs.iterdir():\n",
    "                dir_panel_mafs = dir_panel_maf_dirs / dir_wg_mafs.name\n",
    "                if not dir_panel_mafs.exists():\n",
    "                        dir_panel_mafs.mkdir()\n",
    "                convert_wg_mafs_to_panel_mafs(dir_wg_mafs, panel_df, dir_panel_mafs)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n",
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n",
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n",
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n",
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n",
      "/var/folders/8l/pbcwnq316rz5v5pnyv_sydmn2fsdlk/T/ipykernel_40223/3528665301.py:9: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(filepath, sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "panel_maf_dirs_from_wgs_maf_dirs(dir_maf_dirs, msk_impact_df, dir_panel_maf_dirs)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "outputs": [],
   "source": [
    "def get_sbs_trinucleotide_contexts() -> List[str]:\n",
    "    \"\"\"\n",
    "    Returns a list of trinucleotide context for single base substitutions (SBS)\n",
    "    for constructing a COSMIC mutational spectra matrix.\n",
    "\n",
    "    :return: a list of SBS trinucleotide contexts.\n",
    "    \"\"\"\n",
    "    sbs_trinucleotide_contexts = []\n",
    "    nucleotide_bases = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    substitution_types = [\"C>A\", \"C>G\", \"C>T\", \"T>A\", \"T>C\", \"T>G\"]\n",
    "\n",
    "    for base_5 in nucleotide_bases:\n",
    "        for substitution in substitution_types:\n",
    "            for base_3 in nucleotide_bases:\n",
    "                sbs_trinucleotide_contexts.append(f\"{base_5}[{substitution}]{base_3}\")\n",
    "\n",
    "    return sbs_trinucleotide_contexts\n",
    "\n",
    "\n",
    "def init_sbs_mutational_spectra(n_records: int) -> OrderedDict[str, List[int]]:\n",
    "    \"\"\"\n",
    "    Initilizes an ordered dictionary with SBS trinucleotide context as keys and\n",
    "    a list of counts, one for each sample.\n",
    "\n",
    "    :param n_records: number of samples to record in the mutational spectra matrix.\n",
    "    :return: an ordered dictionary of trinucleotide context and a list of counts\n",
    "        initialized to zeros.\n",
    "    \"\"\"\n",
    "    sbs_mutational_spectra = OrderedDict()\n",
    "    sbs_trinucleotide_contexts = get_sbs_trinucleotide_contexts()\n",
    "\n",
    "    for context in sbs_trinucleotide_contexts:\n",
    "        sbs_mutational_spectra[context] = [0]*n_records\n",
    "\n",
    "    return sbs_mutational_spectra\n",
    "\n",
    "\n",
    "def index_reference_genome(ref_fasta_filepath: Path) -> pyfaidx.Fasta:\n",
    "    \"\"\"\n",
    "    Returns an indexed FASTA file to quickly lookup subsequences in a genome.\n",
    "\n",
    "    :param ref_fasta_filepath: filepath of the FASTA file of the reference genome.\n",
    "    :return: an indexed FASTA file\n",
    "    \"\"\"\n",
    "    return pyfaidx.Fasta(ref_fasta_filepath.as_posix())\n",
    "\n",
    "\n",
    "def get_trinucleotide_ref_from_fasta(row: pd.Series,\n",
    "                                     ref_fasta: pyfaidx.Fasta) -> str:\n",
    "    \"\"\"\n",
    "    Returns the trinucleotides (5' base, reference allele, 3' base) around the\n",
    "    mutation described by the row.\n",
    "\n",
    "    :param row: a pandas row of the MAF file.\n",
    "    :param ref_fasta: an indexed FASTA of the reference genome.\n",
    "    :return: trinucleotide context for the mutation described by the row.\n",
    "    \"\"\"\n",
    "    pointer = int(row[\"chromosome_start\"])\n",
    "    \"\"\"\n",
    "    '-2' and not '-1' because genomes are indexed starting from 1 but Python data\n",
    "    structures are indexed starting from 0.\n",
    "    \"\"\"\n",
    "    return ref_fasta[f\"chr{row['chromosome']}\"][(pointer-2):(pointer+1)].seq.upper()\n",
    "\n",
    "\n",
    "def standardize_trinucleotide(trinucleotide_ref: str) -> str:\n",
    "    \"\"\"\n",
    "    COSMIC signatures define mutations from a pyrimidine allele (C, T) to any\n",
    "    other base (C>A, C>G, C>T, T>A, T>C, T>G). If a mutation in the MAF file\n",
    "    is defined from a purine allele (A, G), then we infer the trinucleotide\n",
    "    context in the complementary sequence, which would be from a pyrimidine\n",
    "    allele due to purines and pyrimidines complementing each other in a\n",
    "    double-stranded DNA.\n",
    "\n",
    "    :param trinucleotide_ref: trinucleotide sequence seen in the reference genome.\n",
    "    :return: a pyrimidine-centric trinucleotide sequence.\n",
    "    \"\"\"\n",
    "    complement_seq = {\n",
    "        'A': 'T',\n",
    "        'C': 'G',\n",
    "        'T': 'A',\n",
    "        'G': 'C'\n",
    "    }\n",
    "    purines = [\"A\", \"G\"]\n",
    "    if trinucleotide_ref[1] in purines:\n",
    "        return f\"{complement_seq[trinucleotide_ref[2]]}\" \\\n",
    "               f\"{complement_seq[trinucleotide_ref[1]]}\" \\\n",
    "               f\"{complement_seq[trinucleotide_ref[0]]}\"\n",
    "    else:\n",
    "        return trinucleotide_ref\n",
    "\n",
    "\n",
    "def standardize_substitution(ref_allele: str,\n",
    "                             mut_allele: str) -> str:\n",
    "    \"\"\"\n",
    "    COSMIC signatures define mutations from a pyrimidine allele (C, T) to any\n",
    "    other base (C>A, C>G, C>T, T>A, T>C, T>G). If a mutation in the MAF file\n",
    "    is defined from a reference purine allele (A, G), then we infer the substituted\n",
    "    base in the complementary sequence, which would be from a pyrimidine\n",
    "    allele due to purines and pyrimidines complementing each other in a\n",
    "    double-stranded DNA.\n",
    "\n",
    "    :param ref_allele: base in the reference genome.\n",
    "    :param mut_allele: base in the mutated genome\n",
    "    :return: substitution string from pyrimidine to any other base.\n",
    "    \"\"\"\n",
    "    complement_seq = {\n",
    "        'A': 'T',\n",
    "        'C': 'G',\n",
    "        'T': 'A',\n",
    "        'G': 'C'\n",
    "    }\n",
    "    purines = [\"A\", \"G\"]\n",
    "    if ref_allele in purines:\n",
    "        return f\"{complement_seq[ref_allele]}>{complement_seq[mut_allele]}\"\n",
    "    else:\n",
    "        return f\"{ref_allele}>{mut_allele}\"\n",
    "\n",
    "\n",
    "def add_instance_to_mutational_spectra(maf_df: pd.DataFrame,\n",
    "                                       mutational_spectra: OrderedDict[str, List[int]],\n",
    "                                       ref_fasta: pyfaidx.Fasta,\n",
    "                                       index: int) -> None:\n",
    "    \"\"\"\n",
    "    Parses each row in a MAF dataframe generated from an ICGC SSM dataset and tabulates a\n",
    "    mutational spectra count matrix in the form of an ordered dictionary.\n",
    "\n",
    "    :param maf_df: MAF dataframe generated from an ICGC SSM dataset.\n",
    "    :param mutational_spectra: an ordered dictionary to tabulat the mutational spectra matrix.\n",
    "    :param ref_fasta: an indexed reference genome.\n",
    "    :param index: row index in the mutational spectra matrix to tabulate in the counts.\n",
    "    \"\"\"\n",
    "    nucleotide_bases = [\"A\", \"C\", \"G\", \"T\"]\n",
    "    pyrimidine = [\"C\", \"T\"]\n",
    "\n",
    "    for _, row in maf_df.iterrows():\n",
    "        if((row[\"chromosome_start\"] != row[\"chromosome_end\"]) or\n",
    "                (row[\"reference_genome_allele\"] not in nucleotide_bases) or\n",
    "                (row[\"mutated_to_allele\"] not in nucleotide_bases)):\n",
    "            continue\n",
    "        trinucleotide_ref = standardize_trinucleotide(\n",
    "            get_trinucleotide_ref_from_fasta(row, ref_fasta))\n",
    "        substitution = standardize_substitution(row[\"reference_genome_allele\"],\n",
    "                                                row[\"mutated_to_allele\"])\n",
    "\n",
    "        # sanity checks\n",
    "        try:\n",
    "            assert (trinucleotide_ref is not None)\n",
    "            assert (trinucleotide_ref[1] == substitution[0])\n",
    "            assert (trinucleotide_ref[1] in pyrimidine)\n",
    "            assert (substitution[0] in pyrimidine)\n",
    "        except AssertionError:\n",
    "            print(f\"MAF row: {row['chromosome']}, \"\n",
    "                  f\"{row['chromosome_start']}, \"\n",
    "                  f\"{row['chromosome_end']}, \"\n",
    "                  f\"{row['reference_genome_allele']}, \"\n",
    "                  f\"{row['mutated_to_allele']}\")\n",
    "            print(f\"FASTA context: {get_trinucleotide_ref_from_fasta(row, ref_fasta)}\")\n",
    "            print(f\"Pyrimidine-centric context: {trinucleotide_ref}\")\n",
    "            raise\n",
    "\n",
    "        mutational_spectra[f\"{trinucleotide_ref[0]}[{substitution}]{trinucleotide_ref[2]}\"][index] += 1\n",
    "\n",
    "\n",
    "def write_mutational_spectra(mutational_spectra: OrderedDict,\n",
    "                             sample_names: List[str],\n",
    "                             filepath: Path) -> None:\n",
    "    \"\"\"\n",
    "    Writes the mutational spectra matrix data, stored in an ordered dictionary, to a CSV file.\n",
    "\n",
    "    :param mutational_spectra: mutational spectra matrix data stored in an ordered dictionary.\n",
    "    :param sample_names: a list of names of the samples.\n",
    "    :param filepath: name of the CSV file to save the data.\n",
    "    \"\"\"\n",
    "    data = np.stack([np.array(mutational_spectra[substitution]) for substitution in mutational_spectra.keys()])\n",
    "    index = pd.Series(\n",
    "        data=mutational_spectra.keys(),\n",
    "        name=\"Mutation Types\"\n",
    "    )\n",
    "    mutational_spectra_df = pd.DataFrame(\n",
    "        data=data,\n",
    "        index=index,\n",
    "        columns=sample_names,\n",
    "        dtype=int,\n",
    "    )\n",
    "    mutational_spectra_df.to_csv(filepath, sep=\",\", index=True)\n",
    "\n",
    "\n",
    "def convert_mafs_to_sbs_mutational_spectra(dir_mafs: Path,\n",
    "                                           ref_fasta_filepath: Path,\n",
    "                                           filepath_output: Path) -> None:\n",
    "    \"\"\"\n",
    "    Converts all MAF files (one file per sample) in a directory into a mutational spectra\n",
    "    matrix and saves it as a CSV file.\n",
    "\n",
    "    :param dir_mafs: a directory containing MAF files.\n",
    "    :param ref_fasta_filepath: filepath to the reference genome FASTA file.\n",
    "    :param filepath_output: file path to save the mutational spectra CSV file.\n",
    "    \"\"\"\n",
    "    maf_filepaths = natsorted(list(dir_mafs.glob(\"*\")))\n",
    "    n_samples = len(maf_filepaths)\n",
    "    mutational_spectra = init_sbs_mutational_spectra(n_samples)\n",
    "    ref_fasta = index_reference_genome(ref_fasta_filepath)\n",
    "    donors = list()\n",
    "\n",
    "    donor_index = 0\n",
    "    for maf_filepath in maf_filepaths:\n",
    "        data_maf = read_sbs_maf_file(maf_filepath)\n",
    "        add_instance_to_mutational_spectra(data_maf, mutational_spectra, ref_fasta, donor_index)\n",
    "        donors.append(maf_filepath.name)\n",
    "        donor_index += 1\n",
    "    write_mutational_spectra(mutational_spectra, donors, filepath_output)\n",
    "\n",
    "\n",
    "def convert_maf_dirs_to_sbs_mutational_spectra(dir_maf_dirs: Path,\n",
    "                                               ref_fasta_filepath: Path,\n",
    "                                               dir_output: Path) -> None:\n",
    "    \"\"\"\n",
    "    For each directory within the specified directory, this method iterates through all\n",
    "    MAF files and creates a mutational spectra matrix and saves them as a CSV file.\n",
    "\n",
    "    :param dir_maf_dirs: a directory of directories, each containing a set of MAF files.\n",
    "    :param ref_fasta_filepath: filepath to the reference genome FASTA file.\n",
    "    :param dir_output: directory to save the mutational spectra CSV files.\n",
    "    \"\"\"\n",
    "    for dir_mafs in dir_maf_dirs.iterdir():\n",
    "        if dir_mafs.is_dir():\n",
    "            filepath_output = dir_output / f\"{dir_mafs.name}.csv\"\n",
    "            convert_mafs_to_sbs_mutational_spectra(dir_mafs, ref_fasta_filepath, filepath_output)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "outputs": [],
   "source": [
    "fpath_grch37 = dir_data / \"hg19.fa\"\n",
    "dir_panel_spectra = dir_data / \"mutational_spectra_panel\"\n",
    "if not dir_panel_spectra.exists():\n",
    "    dir_panel_spectra.mkdir()\n",
    "\n",
    "convert_maf_dirs_to_sbs_mutational_spectra(dir_panel_maf_dirs, fpath_grch37, dir_panel_spectra)"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}